li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -260,27 +296,38 @@ synchronized public void assign(Collection<TopicPartition> partitions) {
       throw new IllegalArgumentException("Topic partition collection to assign to cannot be null");
     }
 
-    // if partitions are empty, it should be treated the same as unsubscribe(). This
-    // is the vanilla Kafka consumer behavior.
+    // If partitions are empty, it should be treated the same as unsubscribe(). This is the vanilla Kafka consumer
+    // behavior.
     if (partitions.isEmpty()) {
       unsubscribe();
       return;
     }
 
-    Map<ClusterDescriptor, Collection<TopicPartition>> clusterToTopicPartitionsMap = getPartitionsByCluster(partitions);
+    // TODO: Throw an exception if there exists a partition whose topic exists but the specified partition number is
+    // out of range for that topic. Whether of not to throw an exception can be controlled by a property.
+    Set<TopicPartition> partitionsSet = new HashSet<>(partitions);
+    PartitionsByClusterResult partitionsByClusterResult = getPartitionsByCluster(partitionsSet);
+
+    _currentSubscription = new ManuallyAssigned(partitionsSet, partitionsByClusterResult.getNonexistentTopics());
+    if (!partitionsByClusterResult.getNonexistentTopics().isEmpty()) {
+      LOG.warn("the following topics in the requested assignment currently do not exist: {}",
+          partitionsByClusterResult.getNonexistentTopics());
+    }
 
     // This assignment should replace the previous assignment. Since max.poll.records should be reset for consumers
     // for the new assignment, close all existing consumers first.
-    List<ClusterConsumerPair<K, V>> consumers = _consumers;
-    if (!consumers.isEmpty()) {
+    if (!_consumers.isEmpty()) {
```

Comment:
in theory, if the set of clusters we talk to (hence the number of consumers) didnt change - we could avoide closing these, right ?

Reply:
Right. I am closing the existing consumers because the number of clusters to talk to for the new assignment may change.

A simpler and probably less bug-prone (but less ideal) approach would be to pre-create consumers for all clusters in the cluster group and fix max.poll.records based on that. Maybe we should do that? We probably don't need clusters groups with more than two clusters until NG takes over. What do you think?

This will require the new Mario query I asked before which will return all clusters in a cluster group.

Reply:
i hate the pre-creation idea. especially as the number of clusters may dynamically change